Mathlib/RingTheory/Polynomial/DegreeLT.lean

```diff
@@ -0,0 +1,238 @@
+  ((degreeLT.basis R m).prod (degreeLT.basis R n)).reindex finSumFinEquiv
+
+@[simp] lemma degreeLT.basis_prod_castAdd (m n : ℕ) (i : Fin m) :
+    basis_prod R m n (i.castAdd n) = (⟨X ^ i.val, degreeLT_X_pow_mem i⟩, 0) := by
+  rw [basis_prod, Basis.reindex_apply, finSumFinEquiv_symm_apply_castAdd]
+  ext
+  · rw [Basis.prod_apply_inl_fst, basis_apply]
+  · rw [Basis.prod_apply_inl_snd]
+
+@[simp] lemma degreeLT.basis_prod_natAdd (m n : ℕ) (i : Fin n) :
+    basis_prod R m n (i.natAdd m) = (0, ⟨X ^ i.val, degreeLT_X_pow_mem i⟩) := by
+  rw [basis_prod, Basis.reindex_apply, finSumFinEquiv_symm_apply_natAdd]
+  ext
+  · rw [Basis.prod_apply_inr_fst]
+  · rw [Basis.prod_apply_inr_snd, basis_apply]
+
+/-- An isomorphism between `(R[x]_(n+m))` and `(R[x]_n) × (R[x]_m)` given by the fact that
+the bases are both indexed by `Fin (n+m)`. -/
+noncomputable def degreeLT.addLinearEquiv {n m : ℕ} :
+    (R[x]_(n + m)) ≃ₗ[R] (R[x]_n) × (R[x]_m) :=
+  Basis.equiv (degreeLT.basis _ _) (degreeLT.basis_prod _ _ _) (Equiv.refl _)
+
+@[aesop unsafe 50%]
+theorem degree_add_lt_of_degree_lt {p q : R[X]} {n : ℕ}
+    (hp : degree p < n) (hq : degree q < n) :
+    degree (p + q) < n :=
+  (degree_add_le p q).trans_lt <| max_lt hp hq
+
+@[aesop unsafe 50%]
+theorem degree_mul_lt_of_lt_left {p q : R[X]} {a : WithBot ℕ} {b : ℕ}
+    (hp : degree p < a) (hq : degree q ≤ b) :
+    degree (p * q) < a + b := by
+  by_cases hq0 : q.degree = ⊥
+  · rw [degree_eq_bot] at hq0
+    rw [hq0, mul_zero, degree_zero, WithBot.bot_lt_add]
+    simp [WithBot.bot_lt_iff_ne_bot, ne_bot_of_gt hp]
+  · exact (degree_mul_le _ _).trans_lt (WithBot.add_lt_add_of_lt_of_le hq0 hp hq)
+
+@[aesop unsafe 50%]
+theorem degree_mul_lt_of_lt_right {p q : R[X]} {a : ℕ} {b : WithBot ℕ}
+    (hp : degree p ≤ a) (hq : degree q < b) :
+    degree (p * q) < a + b := by
+  by_cases hp0 : p.degree = ⊥
+  · rw [degree_eq_bot] at hp0
+    rw [hp0, zero_mul, degree_zero, WithBot.bot_lt_add]
+    simp [WithBot.bot_lt_iff_ne_bot, ne_bot_of_gt hq]
+  · exact (degree_mul_le _ _).trans_lt (WithBot.add_lt_add_of_le_of_lt hp0 hp hq)
+
+lemma mul_left_mem_degreeLT {n} (p q : R[X]) (hp : degree p ≤ m) (hq : q ∈ R[x]_n) :
+    p * q ∈ R[x]_(m + n) := by
+  rw [mem_degreeLT] at *
+  exact degree_mul_lt_of_lt_right hp hq
+
+lemma mul_left_mem_degreeLT' {n} (p q : R[X]) (hp : degree p ≤ m) (hq : q ∈ R[x]_n) :
+    p * q ∈ R[x]_(n + m) := by
+  rw [add_comm]
+  exact mul_left_mem_degreeLT _ _ ‹_› ‹_›
+
+lemma mul_right_mem_degreeLT {m} (p q : R[X]) (hp : p ∈ R[x]_m) (hq : degree q ≤ n) :
+    p * q ∈ R[x]_(m + n) := by
+  rw [mem_degreeLT] at *
+  exact degree_mul_lt_of_lt_left hp hq
+
+lemma mul_right_mem_degreeLT' {m} (p q : R[X]) (hp : p ∈ R[x]_m) (hq : degree q ≤ n) :
+    p * q ∈ R[x]_(n + m) := by
+  rw [add_comm]
+  exact mul_right_mem_degreeLT _ _ ‹_› ‹_›
+
+end degreeLT
+
+section translate
+
+variable {R : Type*} [CommRing R] {x : R} {m n : ℕ} {a : R} {f g : R[X]}
+
+variable (x) in
```

Comment:
I would use variable `a` not `x` for this everywhere, I think x is too close to X. And I would put it explicitly in the def in the code i.e. `translate a : R[X] ≃+* R[X]` so people can see it. And are you definitely sure about the minus sign? And yes, I think that this could easily go in a file of its own. My instinct is that the more files the better, and this is really not about DegreeLT at all. But I'd be happy to see the DegreeLT file kept, the Basic file is getting too big anyway.